src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Configuration
+{
+    public class IncludeConfigurationProvider : ConfigurationProvider
+    {
+        public IncludeConfigurationProvider(IConfiguration source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            foreach (var section in source.GetChildren())
+            {
+                AddSection(section, section.Key);
+            }
+        }
+
+        private void AddSection(IConfigurationSection section, string keyPrefix)
+        {
+            Data.Add(keyPrefix, section.Value);
+            foreach (var child in section.GetChildren())
+            {
+                AddSection(child, keyPrefix + Constants.KeyDelimiter + child.Key);
```

Comment:
Yeah good idea, that seems to work.  The cast for the root smells a bit.  I think this is another example where it'd be nicer to have Path/Key in the IConfiguration and simply handle the Root to have empty/null Path/Key.  Doing is casts/ checks in traversal code is pretty gross.  I updated the PR moving Path/key to the base and having them return "".  I am willing to bet you are going to have some pushback but its worth a shot :)